CRM/Civicase/Hook/BuildForm/TokenTree.php

```diff
@@ -0,0 +1,392 @@
+<?php
+
+/**
+ * Attaches a new token tree to the form.
+ */
+class CRM_Civicase_Hook_BuildForm_TokenTree {
+
+  const CASE_ROLE_TOKEN_TEXT = 'Case Roles';
+
+  const CURRENT_USER_TOKEN_TEXT = 'Current User';
+
+  const CASE_TOKEN_TEXT = 'Case';
+
+  const CONTACT_TOKEN_TEXT = 'Contact';
+
+  const CLIENT_TOKEN_TEXT = 'Client';
+
+  const OTHER_TOKEN_TEXT = 'Other';
+
+  const ADDRESS_TOKEN_TEXT = 'Address';
+
+  /**
+   * All case and contact related custom fields.
+   *
+   * @var array
+   */
+  private $customFields = [];
+
+  /**
+   * Attaches a new token tree to the form.
+   *
+   * @param CRM_Core_Form $form
+   *   Form object class.
+   * @param string $formName
+   *   Form name.
+   */
+  public function run(CRM_Core_Form &$form, $formName) {
+    if (!$this->shouldRun($formName)) {
+      return;
+    }
+    $this->fetchAllRelevantCustomFields();
+    $this->attachNewTokenTreeToForm($form);
+  }
+
+  /**
+   * Attaches a new token tree to the form.
+   *
+   * @param CRM_Core_Form $form
+   *   Form object class.
+   */
+  private function attachNewTokenTreeToForm(CRM_Core_Form &$form) {
+    $tokens = $form->get_template_vars('tokens');
+    $newTokenTree = [];
+    foreach ($tokens as $key => $tokenList) {
+      if (!empty($tokenList['text']) && $tokenList['text'] === self::CASE_ROLE_TOKEN_TEXT) {
+        $this->addCaseRoleTokens($tokenList['children'], $newTokenTree);
+      }
+      elseif (!empty($tokenList['text']) && $tokenList['text'] === self::CURRENT_USER_TOKEN_TEXT) {
+        $this->addCurrentUserTokens($tokenList['children'], $newTokenTree);
+      }
+      elseif (!empty($tokenList['text']) && $tokenList['text'] === self::CASE_TOKEN_TEXT) {
+        $this->addCaseTokens($tokenList['children'], $newTokenTree);
+      }
+      elseif (!empty($tokenList['text']) && in_array(
+        $tokenList['text'],
+        [self::CONTACT_TOKEN_TEXT, self::ADDRESS_TOKEN_TEXT]
+        )) {
+        $this->addClientTokens($tokenList['children'], $newTokenTree);
+      }
+      else {
+        $this->addOtherTokens($tokenList['children'], $newTokenTree);
+      }
+    }
+    $this->reFormatCustomTokens($newTokenTree);
+    CRM_Core_Resources::singleton()
```

Comment:
Can we move this to a new private function `addTokenToJS` or something better.

Reply:
done